### PR TITLE
pg_compression: add Zstandard

### PR DESCRIPTION
**Pull request by leskin-in**
As described in [this issue](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fgreenplum-db%2Fgpdb%2Fissues%2F3155&sa=D&sntz=1&usg=AFQjCNHitXVIaj8vyBb2H32TkPAqUSSC2w) and in [GPDB mailing list topic](https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/XoYcF97qK-Y), there is a need of some new compression algorithm in GPDB.

We (@leskin-in @kapustor, @acmnu) have added integration with the [Zstandard](https://github.com/facebook/zstd) library.

A simple test conducted on CentOS 7 one-core machine shows that `zstd` is approximately 4 times faster than `zlib`, while compressed table size is almost the same for them both. We used a compression level 9 and a raw (uncompressed) table size was 248Mb. Here is a script:
```
psql -d postgres -c 'CREATE TABLE sales_zstd (txn_id int, qty text) WITH (appendonly=true, compresstype=zstd, compresslevel=9) DISTRIBUTED BY (txn_id);'
psql -d postgres -c 'CREATE TABLE sales_zlib (txn_id int, qty text) WITH (appendonly=true, compresstype=zlib, compresslevel=9) DISTRIBUTED BY (txn_id);'
psql -d postgres -c 'CREATE TABLE sales_none (txn_id int, qty text) WITH (appendonly=true, compresstype=none) DISTRIBUTED BY (txn_id);'

time psql -d postgres -c 'insert into sales_zstd (select generate_series(1,10000000) as txn_id, generate_series(1,10000000)::text as qty);'
# real	0m19.479s
time psql -d postgres -c 'insert into sales_zlib (select generate_series(1,10000000) as txn_id, generate_series(1,10000000)::text as qty);'
# real	1m37.831s
time psql -d postgres -c 'insert into sales_none (select generate_series(1,10000000) as txn_id, generate_series(1,10000000)::text as qty);'
# real	0m18.173s

psql -d postgres -c 'SELECT * FROM pg_size_pretty(pg_total_relation_size('sales_zstd'));'
# 44 MB
psql -d postgres -c 'SELECT * FROM pg_size_pretty(pg_total_relation_size('sales_zlib'));'
# 52 MB
psql -d postgres -c 'SELECT * FROM pg_size_pretty(pg_total_relation_size('sales_none'));'
# 248 MB
```

However, there are a few issues with this solution:
1. Zstandard supports 19 (safe, see http://facebook.github.io/zstd/zstd_manual.html#Chapter1) compression levels. As far as I understand, the current version of `src/backend/utils/datumstream.c` does not allow to use compression levels >= 10 for some reason
2. It is unclear how to disable compilation and linking of the file `zstd_compression.c` if the flag `--with-zstd` was not passed to `configure` script (at the moment, `zstd` is disabled by default). The Make command for that is in `src/backend/catalog/Makefile`. This is a reason why the CI tests cannot be passed

We will be glad to receive feedback from the community. As for those issues, we are going to fix them in a few days.

### CODE REVIEW

**Comment by Eulerizeit** [comment]
@schubert Zstandard is licensed under both BSD and GPLv2 My understanding is that this isn't an issue but I would like to double check.

File: configure [hunk 1]
```diff
@@ -1604,6 +1618,7 @@ Optional Packages:
                           use system time zone data in DIR
   --without-zlib          do not use Zlib
   --without-libbz2        do not use bzip2
+  --with-zstd             do not use zstd
```
Comment (tvesely) [inline configure]: The caption here is backward. It should be changed to something like "build with zstd support"

**Comment by hlinnaka** [comment]
Great, thanks! I'll have a closer look later, but this looks good at a quick glance.

**Comment by hlinnaka** [comment]
> 1. Zstandard supports 19 (safe, see http://facebook.github.io/zstd/zstd_manual.html#Chapter1) compression levels. As far as I understand, the current version of src/backend/utils/datumstream.c does not allow to use compression levels >= 10 for some reason

I think it's just because zlib uses levels 1-9. Perhaps we need to resurrect or reinvent the "validation" functions Or just always accept 1-19, and just treat values higher than 9 the same as 9, for zlib.

> 2. It is unclear how to disable compilation and linking of the file zstd_compression.c if the flag --with-zstd was not passed to configure script (at the moment, zstd is disabled by default). The Make command for that is in src/backend/catalog/Makefile. This is a reason why the CI tests cannot be passed

Hmm. might be easiest to always compile the `zstd_compression.c` source file, but put the contents inside a huge `#ifdef HAVE_LIBZSTD` that covers the whole file. And add an `#else` branch that contains dummy versions of the functions that just throw an error. Or put #ifdef to the individual functions.

If you want to omit the file completely, you'll need to have something like this in the Makefile:
```
ifeq ($(with_zstd),yes)
OBJS += zstd_compression.o
endif
```
But then you'll need to provide the dummy versions of the functions in another file. Or have the pg_proc entries for them inside #ifdefs as well, which might be a bit tricky.

**Comment by ashwinstar** [comment]
Great work and nice progress. Adding tests for functionality in ICW with the PR would be good.

**Comment by hlinnaka** [comment]
Ok, I spent some time massaging this into a committable state:

* a lot of little cleanup of whitespace, error messages, etc.
* removed the function pointers for compression/decompression functions, they weren't really used for anything that'd need the indirection.
* added a regression test
* updated the documentation (I didn't actually build the docs, so it's likely broken)

I think this is now in good shape, and can be committed. I plan to squash all of the commits into one before pushing, but I kept the history for now, for easier review. For the final commit message, I propose:
```
Author: Ivan Leskin <leskin.in@arenadata.io>

    Add Zstandard compression option for append-optimized tables
    
    Add a new compression option for append-optimized tables, "zstd". It is generally
    faster than zlib or quicklz, and compresses better. Or at least it can be faster or
    compress better, if not both at the same time, by adjusting the compression level.
    A major advantage of Zstandard is the wide range of tuning, to choose the trade-off
    between compression speed and ratio.

    Update documentation to mention "zstd" alongside "zlib" and "quicklz". More could
    be done; all the examples still use zlib or quicklz, for example, and I think we want
    to emphasize Zstandard more in the docs, over those other options. But this is the
    bare minimum to keep the docs factually correct.
    
    Using the new option requires building the server with the libzstd library. A new
    --with-zstd option is added for that. The default is to build without libzstd, for now,
    but we should probably change the default to be on, after we have had a chance
    to update all the buildfarm machines to have libzstd.
```

Please have a final look-over these changes. Barring objections or concourse pipeline failures, I'll squash and push this later tonight or on Monday.

**Comment by leskin-in** [comment]
Great job, @hlinnaka! I really appreciate your help.
It's now more clear to me how to contribute to GPDB, too.

**Comment by hlinnaka** [comment]
Ok, squashed and pushed as commit 724f9d27b160ec3f87a5ce95eecc605808be3ced.

Many thanks for the patch! This is really important, because in the long run, this allows us to deprecate zlib and QuickLZ. Maybe even the RLE compression, if the performance of this is good enough.

@pivotal-mike, we're going to need to change the configuration of the build machines on the pipeline, to build with `--with-zstd`. And for that, libzstd needs to be enabled.

@mkiyama, I made minimal docs changes, to list "zstd" alongside "zlib" in the reference pages. But I think we want more changes, and I think we should start recommending zstd over quicklz or zlib.


Also, I think we should switch to using zstd for compressing the plan trees, when they're dispatched from QD to QEs. That could really benefit from the faster compression/decompression speed. Also, should off "zstd" as an option for workfile compression (`gp_workfile_compress_algorithm`), for the sake of completeness.

**Comment by ashwinstar** [comment]
Don't see tests added for CO tables. Also, for CO tables good to evaluate if having zstd as block compression instead of current RLE + zlib is better.

**Comment by hlinnaka** [comment]
On 01/12/17 21:46, Ashwin Agrawal wrote:
> Don't see tests added for CO tables. Also, for CO tables good to evaluate if having zstd as block compression instead of current RLE + zlib is better.

There is this in the new test:
```
+CREATE TABLE zstdtest (id int4, t text) WITH (appendonly=true, 
compresstype=zstd, orientation=column);
+
+INSERT INTO zstdtest SELECT g, 'foo' || g FROM generate_series(1, 
100000) g;
+INSERT INTO zstdtest SELECT g, 'bar' || g FROM generate_series(1, 
100000) g;
...
```

It's not much, but I think that's enough. As far as the compression code 
is concerned, it doesn't really make any difference whether the table is 
row- or column-oriented.

Yeah, some performance testing would be nice, so that we know what we 
should recommend and/or deprecate going forward.

- Heikki

**Pull request closed without merging**